app/models/reports/repository.rb

```diff
@@ -1,64 +1,143 @@
 module Reports
   class Repository
+    include Memery
+    PERCENTAGE_PRECISION = 0
+
     def initialize(regions, periods:, with_exclusions: false)
       @regions = Array(regions)
-      @regions_by_id = @regions.group_by { |r| r.id }
       @with_exclusions = with_exclusions
 
       @periods = if periods.is_a?(Period)
-        Range.new(periods, periods).to_a
+        Range.new(periods, periods)
       else
-        periods.to_a
+        periods
       end
     end
 
-    attr_reader :regions, :periods
-    attr_reader :regions_by_id
+    attr_reader :periods
+    attr_reader :regions
     attr_reader :with_exclusions
 
     delegate :cache, :logger, to: Rails
 
-    def controlled_patients_count
-      cached_query(:controlled_patients_count) do |item|
-        control_rate_query.controlled(item.region, item.period, with_exclusions: with_exclusions).count
+    # Uses Memery to memoize a method, but takes into account our force_cache setting. If force_cache is true,
+    # a caller is asking for all values to be retrieved fresh from the database, so we want to skip memoization.
+    def self.smart_memoize(method)
+      memoize(method, condition: -> { !force_cache? })
+    end
+
+    # Returns assigned patients for a Region. NOTE: We grab and cache ALL the counts for a particular region with one SQL query
+    # because it is easier and fast enough to do so. We still return _just_ the periods the Repository was created with
+    # to conform to the same interface as all the other queries here.
+
+    # Returns a Hash in the shape of:
+    # {
+    #    region_slug: { period: value, period: value },
+    #    region_slug: { period: value, period: value }
+    # }
+    smart_memoize def assigned_patients_count
+      full_assigned_patients_counts.each_with_object({}) do |(entry, result), results|
```

Comment:
Is it possible to do a `group_by_period` and fill in `nils` after? That might be faster.

Reply:
not sure I follow - this is grabbing just the subset of assigned patient counts the caller is asking for here.